src/Framework/AspNetCoreAnalyzers/src/SourceGenerators/PublicTopLevelProgramGenerator.cs

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.AspNetCore.SourceGenerators;
+
+[Generator]
+public class PublicProgramSourceGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var internalGeneratedProgramClass = context.CompilationProvider.Select((compilation, cancellationToken) =>
+        {
+            var program = compilation.GetTypeByMetadataName("Program");
+            // If the program class is already public, we don't need to generate anything.
+            if (program is null || program.DeclaredAccessibility == Accessibility.Public)
+            {
+                return null;
+            }
+            // If the discovered `Program` type is an interface, struct or generic type then its not
+            // generated and has been defined in source, so we can skip it
+            if (program.TypeKind == TypeKind.Struct || program.TypeKind == TypeKind.Interface || program.IsGenericType)
```

Comment:
Out of curiosity, why not `program.TypeKind != TypeKind.Class`?

Reply:
That's....a good idea 😅  I'll update as needed.